5calls/app/src/androidTest/java/org/a5calls/android/a5calls/controller/MainActivityErrorTest.java

```diff
@@ -0,0 +1,199 @@
+        FakeRequestQueue requestQueue = new FakeRequestQueue(basicNetwork);
+        requestQueue.start();
+
+        // Replace the app's RequestQueue with our mock
+        AppSingleton.getInstance(
+                InstrumentationRegistry.getInstrumentation().getTargetContext())
+                .setRequestQueue(requestQueue);
+
+        // Create a new FiveCallsApi with our mock RequestQueue
+        String callerId = AccountManager.Instance.getCallerID(
+                InstrumentationRegistry.getInstrumentation().getTargetContext());
+        FiveCallsApi api = new FiveCallsApi(callerId, requestQueue);
+        AppSingleton.getInstance(
+                InstrumentationRegistry.getInstrumentation().getTargetContext())
+                .setFiveCallsApi(api);
+    }
+
+    /**
+     * Launches the MainActivity and waits for it to load
+     * @param waitTimeMs time to wait for the activity to load and process errors
+     */
+    private void launchMainActivity(int waitTimeMs) {
+        // Launch the activity
+        scenario = ActivityScenario.launch(MainActivity.class);
+
+        // Wait for error processing and UI to update
+        try {
+            Thread.sleep(waitTimeMs);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * Custom matcher to check if a RecyclerView has at least one item
+     */
+    public static Matcher<View> hasAtLeastOneItem() {
+        return new TypeSafeMatcher<View>() {
+            @Override
+            protected boolean matchesSafely(View view) {
+                if (!(view instanceof RecyclerView)) {
+                    return false;
+                }
+                RecyclerView recyclerView = (RecyclerView) view;
+                return recyclerView.getAdapter() != null && recyclerView.getAdapter().getItemCount() > 0;
+            }
+
+            @Override
+            public void describeTo(Description description) {
+                description.appendText("RecyclerView with at least one item");
+            }
+        };
+    }
+
+    /**
+     * Verifies that error UI is displayed correctly
+     */
+    private void verifyErrorUI() {
+        // Check that the error message is displayed
+        onView(withText(R.string.request_error))
+                .check(matches(isDisplayed()));
+
+        // Verify that the RecyclerView is displayed and has at least one item
+        onView(withId(R.id.issues_recycler_view))
+                .check(matches(isDisplayed()))
+                .check(matches(hasAtLeastOneItem()));
+    }
+
+    @Test
+    public void testNetworkErrorDisplaysSnackbar() {
+        // Set up mock to throw network error
+        mHttpStack.setExceptionToThrow(new IOException("Network error"));
+
+        // Set up mock request queue
+        setupMockRequestQueue();
+
+        // Launch the activity and wait for error processing
+        launchMainActivity(3000);
```

Comment:
this is way more readable now. thanks!

it would be great to eventually figure out how to wait until the UI we want shows up, but I'm OK with thread.sleep indefinitely. It's not like this'll be run on lots of devices automatically -- it'll just be one-offs or at most on github at commit. Could add a TODO so we can someday revisit.

Reply:
added todo